Sign In/Up

pray92 opened the pull request:
회원가입 및 로그인에 필요한 기본적인 기능들을 구현했습니다.

구현사항:
1. 유효성 검사: 아이디, 비밀번호, 닉네임, 이메일
2. 회원가입
3. 로그인

인증 기능은 해당 프로젝트의 기본 기능들을 구현 이후에 작업에 들어가도록 하겠습니다. 추가적으로 구현했으면 좋을 사항이 있으시다면 피드백주시기 바랍니다.

pray92 committed:
CHORE: 서비스 명칭 수정

pray92 committed:
Merge branch 'develop' into 5-feat-sign-up-1

pray92 committed:
Merge branch 'develop' into 5-feat-sign-up-1

pray92 committed:
CHORE: common 패키지 정리

pray92 committed:
CHORE: DEPRECATED code chored

pray92 committed:
FEAT: Implement User Entity

pray92 committed:
CHORE: Delete unused source

pray92 committed:
FEAT: Add Schema based on implemented entity

pray92 committed:
CHORE: CommonExceptionHandler moved to error package

pray92 committed:
CHORE: Setting ddl-auto for each profile yaml

- add simple insert queries for working well

pray92 committed:
CHORE: com -> kr

pray92 committed:
TEST: Test case for overlapped account | nickname | email

pray92 committed:
REFACTOR: let properties in User entity change primitive type

pray92 committed:
FEAT: Implement account id validation REST API

pray92 committed:
FEAT: Implement password validation REST API

pray92 committed:
FEAT: Implement email validation REST API

- Since it depends on validation annotation, test case not included

pray92 committed:
FEAT: Implement nickname validation REST API

- It supports all asian languages

pray92 committed:
CHORE: *ValidRequest to *Request

pray92 committed:
FEAT: Duplication Exception logics

- AccountID
- Nickname
- Email

pray92 committed:
FEAT: Sign up & Sign in

pray92 committed:
CHORE: Add comments, and delete unused code

f-lab-nolan commented on src/main/java/kr/texturized/muus/application/service/UserSignUpService.java:
```diff
@@ -0,0 +1,114 @@
+ */
+@Slf4j
+@Service
+@Transactional
```
Transaction은 클래스에 걸지 않고 보통 메소드 단위에 선언합니다.

pray92 replied:
`signIn()`, `signUp()` 같은 메서드에만 적용하였습니다.

f-lab-nolan commented on src/main/java/kr/texturized/muus/application/service/UserSignUpService.java:
```diff
@@ -0,0 +1,114 @@
+                .userType(UserType.USER)
+            .build()));
+
+        user.ifPresent(u -> log.info("Sign-up: {}", u));
```
https://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/

한번 읽어 보세요

f-lab-nolan replied:
71라인도 같이 해결될거로 보입니다.

pray92 replied:
해당 부분은 블로그 내용을 토대로하여 작성하였습니다.

다만 제가 Optional을 사용하면서 `map()`을 통해 반환 값이 `EMPTY`가 아니면 log를 남기는 방식을 취했는데 보시기에 어떤가요?

f-lab-nolan replied:
일반적으로 사용하는 패턴들이라 괜찮아 보입니다.

f-lab-nolan commented on src/main/java/kr/texturized/muus/presentation/api/UserController.java:
```diff
@@ -1,30 +1,127 @@
+     * @param request for email, it validates using bean validation.
+     * @return Available response.
+     */
+    @ResponseStatus(HttpStatus.ACCEPTED)
```
일반적으로 HttpStatus.OK를 씁니다.
202 Accepted 클라이언트의 요청은 정상적이나, 서버가 아직 요청을 완료하지 못했다의 의미입니다.

f-lab-nolan commented on src/main/java/kr/texturized/muus/presentation/api/UserController.java:
```diff
@@ -1,30 +1,127 @@
+     */
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping("/sign-up")
+    public UserResponse signUp(
```
대개 하나의 DTO로 받는데요..
굳이 이렇게 어렵게 받을 이유가 있을가여?

pray92 replied:
다소 번거롭게 받은 이유는 `재사용성`과 `유지보수`입니다.

아이디, 비밀번호, 닉네임 관련 정책은 유동적으로 바뀔 것 같다는 판단이 섰습니다. 현재 해당 정보들은 `Request` record에서 `@Pattern`을 통해 정책이 처리되고 있습니다. 

여기서 따로 로그인, 회원가입 전용 DTO를 만들고 @Pattern을 적용하면 정책을 작성해야 하는 코드가 늘기 때문에 정책 변경 시 여러 곳을 수정해야할 것이라 예상했습니다. 그래서 따로 DTO를 제작하지 않고 이미 존재하는 request record를 재사용하는 방식을 선택했습니다.

아래에 대한 답변도 포함합니다. @RequestParam 처리로 단순하게 갈 수 있을지도 모르나 Controller 메서드 파라미터에 validation 관련 코드를 추가해야 합니다. 이는 결국 중복된 코드를 작성하게 됩니다. 

또한 @RequestParam 사용 시 클라이언트가 사용한 고유 정보(password)가 노출됩니다. 물론 password만 따로 예외 처리를 둘 수도 있겠지만, 코드의 일관성을 위해 현 상태를 유지하는 것이 좋을 것 같아 의견 내봅니다.

pray92 replied:
해당 사항은 swagger를 통해 다시 확인해보니 클라이언트에서 요청을 전송하는데 어려움이 있어보였습니다. 이를 위해 다시 고민해보고 반영하겠습니다.

f-lab-nolan commented on src/main/java/kr/texturized/muus/presentation/api/UserController.java:
```diff
@@ -1,30 +1,127 @@
+     */
+    @ResponseStatus(HttpStatus.ACCEPTED)
+    @PostMapping("/validate/nickname")
+    public NicknameResponse validateNickname(@RequestBody @Valid final NicknameRequest request) {
```
String으로 선언하고, 받을수 있는 파라미터 명을 명시적으로 선언하면,
굳이 객체로 안받아도 될것 같은데요..

f-lab-nolan replied:
@RequestParam 활용은 안되나요?

f-lab-nolan commented on src/main/java/kr/texturized/muus/presentation/api/UserController.java:
```diff
@@ -1,30 +1,127 @@
+     * @return Available response.
+     */
+    @ResponseStatus(HttpStatus.ACCEPTED)
+    @PostMapping("/validate/account")
```
http method 관련하여 한번 읽어보세요.

https://inpa.tistory.com/entry/WEB-%F0%9F%8C%90-HTTP-%EB%A9%94%EC%84%9C%EB%93%9C-%EC%A2%85%EB%A5%98-%ED%86%B5%EC%8B%A0-%EA%B3%BC%EC%A0%95-%F0%9F%92%AF-%EC%B4%9D%EC%A0%95%EB%A6%AC

pray92 replied:
HttpStatus.ACCEPTED를 OK로 커밋했습니다. 하지만 HttpStatus의 의미를 잘 모르는 것 같습니다. 해당 블로그 글 읽으면서 개념 정리하겠습니다.

f-lab-nolan replied:
위의 거는 http method 관련입니다.
Status 는 아래 관련 보세요.

https://www.whatap.io/ko/blog/40/

pray92 committed:
FIX: Delete @Transactional in Service class

- Add @Transactional sign-in and sign-up only

pray92 committed:
REFACTOR: Sign in and sign up

- Let service focus on returning user and log, don't throw exception while creation excluding validation
- Then let controller focus on parameter validation and throwing exception when service returned empty value

pray92 committed:
REFACTOR: HttpStatus.CREATED -> OK

f-lab-nolan commented on src/main/java/kr/texturized/muus/application/service/UserSignUpService.java:
```diff
@@ -0,0 +1,112 @@
+     * @return user with {@code Optional<T>} wrapper class,
+     *         Optional is recommended to use for return result
+     */
+    @Transactional
```
여기에 Transactional이 필요한 이유가 무엇일가요?

pray92 replied:
현재 signIn() 로직에는 단순 select만 존재하기 때문에 @Transactional(readonly = true)를 통해 DB 읽기 작업에 대해 최적화를 수행하였습니다.

f-lab-nolan commented on src/main/java/kr/texturized/muus/infrastructure/mapper/UserViewMapper.java:
```diff
@@ -1,13 +1,16 @@
+
+    boolean existsByEmail(final String email);
+
+    User findByAccountId(final String accountId);
```
User findByAccountId(final String accountId);
-> Optional<User> findByAccountId(final String accountId);
이렇게도 가능합니다.

pray92 replied:
반영했습니다.

f-lab-nolan commented on src/main/java/kr/texturized/muus/presentation/api/UserController.java:
```diff
@@ -1,30 +1,126 @@
+        return new UserResponse(userSignUpService.signIn(
+            accountRequest.accountId(),
+            passwordRequest.password()
+        ).orElseThrow(InvalidAccountException::new));
```
Service에서 Throw 를 던지는거는 어떨가요?

pray92 replied:
말씀드린 사항대로 service 단에서 throw를 던지도록 처리했습니다.

pray92 committed:
FIX: MyBatis + JPA Mapping, enum mapping and return optional rather than pure entity in myBatis

- Select query by myBatis not worked well for spelling issues, so it fixed
- let enum save in db value set in code rather than ordinal for maintenance issues, and implemented mappers for each mybatis and JPA
- Add test case for UserViewMapper(I found it load improperly, so I added it)

pray92 committed:
CHORE: Add WebMvcConfigurer

- It was coded for get messages for validation from properties.
- But it's hard to get regular expression, so it isn't used.
- Let's leave this for future

pray92 committed:
CHORE: Delete WebMvcConfigurer

- Delete unused sources for now
- Reason for doing this is to record some useful sources

pray92 committed:
CHORE: UserController test accepted -> ok

f-lab-nolan commented on src/main/java/kr/texturized/muus/domain/exception/UserTypeNotFoundException.java:
```diff
@@ -0,0 +1,14 @@
+public class UserTypeNotFoundException extends BusinessException {
+
+    public UserTypeNotFoundException(int value) {
+        super("No matched value: " + value + " in UserType", ErrorCode.INTERNAL_SERVER_ERROR);
```
resource를 못찾는 경우는 보통 400입니다만.
잘못된 타입이 온거라면 400이 일반적이에요.

pray92 replied:
해당 exception이 throw되는 경우는 DB에서 잘못된 UserType 매핑 정수를 가져왔을 경우입니다.

현재 USER 타입을 1로, ADMIN이라는 타입을 100으로 매핑했습니다(이를 위해 JPA(infrastructure/handler) 및 MyBatis(infrastructure/converter) 전용 매퍼 클래스를 작성했습니다). 이렇게 구성한 이유는 유저 타입이 기획에 따라 얼마든지 추가 및 변경될 것이라 생각해서였습니다. 

그래서 이후 기획으로 인해 새로운 유저 타입이 추가된 경우 이에 대한 처리를 코드에 작성해서 처리할 예정입니다. exception은 이에 대한 처리를 하지 않았을 경우 개발 단계에서 exception을 throw하기 위해 작성했습니다.

그래서 이는 서버 내부에서 발생한 문제이기 때문에 500을 던졌습니다. 혹시 이러한 경우여도 400을 던지는 것이 일반적인가요??

f-lab-nolan replied:
일반적으로 앞단에서 제대로 validation을 했다면 정상적인 값들이 있겠죠??
그게 아니라 누군가 DB로 업데이트 하거나요?(이건 해킹이나 내부 범죄자이거나,.)

그리고 비정상 값이라고 오류를 내지는 않아요.
그냥 비정상 값으로 api 결과를 내려줘요...

다만 다른 로직에서는 해당 값으로 처리를 하다가 이상하다면 500에러가 나고 로그를 남겨서
오류 파악을 하지요.

pray92 replied:
비정상적인 값을 받았을 땐 클라이언트 쪽 책임이고, 비즈니스 로직 쪽은 서버 쪽 책임으로 봐야하는거였군요. 반영했습니다.

pray92 committed:
FIX: let sign in logic works in readonly transaction

pray92 committed:
FIX: 500 -> 400 for UserTypeNotFoundException
